### PR TITLE
Fix behavior difference between tf.io.GFile and python file for utf8

### LINKED ISSUE #33563: tf.io.GFIle not working correctly with UTF-8 files and Python3
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu
- TensorFlow installed from (source or binary): source internal Google
- TensorFlow version (use command below): 1.5.0, internal Google
- Python version: 3.6.7

**Describe the current behavior**
Calling 'read(X)' on the text files opened with GFile in python3 doesn't work properly (it fetches X bytes rather than X characters). This often results with the UnicodeDecodeError (as the read can happen in the middle of the unicode character).

**Describe the expected behavior**
It should behave like python3: reading the X characters.

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.

```python
my_text = 'Bären'
with open('/tmp/ex1', 'w') as f:
  f.write(my_text)

// Will read the whole string correctly.
with open('/tmp/ex1', 'r') as f:
  print(f.read())

// This will print 2 chars Ba
with open('/tmp/ex1', 'r') as f:
 print(f.read(2))

// This will print 3 chars: Bar
with open('/tmp/ex1', 'r') as f:
 print(f.read(3))

// This will print the whole thing.
with tf.io.gfile.GFile('/tmp/ex1', 'r') as f:
  print(f.read())

// This will crash.. :-(
with tf.io.gfile.GFile('/tmp/ex1', 'r') as f:
  print(f.read(2))
```

**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

The error will be:
```
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xc3 in position 1: unexpected end of data
```

### PR DESCRIPTION
**Pull request by yongtang**
This PR tries to address the issue raised in #33563 where
tf.io.GFile behavior is different from python file for utf8.

The issue was that tf.io.GFile's read does not take utf8 into
consideration for non binary mode.

This PR fixes the discrepancy.

This PR fixes #33563

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

File: tensorflow/python/lib/io/file_io.i [hunk 1]
```diff
@@ -237,6 +237,67 @@ string ReadFromStream(tensorflow::io::BufferedInputStream* stream,
   return result;
 }
 
+string ReadFromStreamUTF8(tensorflow::io::BufferedInputStream* stream,
+                          size_t bytes,
+                          TF_Status* status) {
+  tensorflow::tstring result;
```
Comment (mihaimaruseac) [inline tensorflow/python/lib/io/file_io.i]: Should this be `std::string` or `tensorflow::tstring`? Afaik, `tensorflow:tstring` should be used only if the string is in a tensor. Adding @gharibian to confirm

File: tensorflow/python/lib/io/file_io.i [hunk 2]
```diff
@@ -237,6 +237,67 @@ string ReadFromStream(tensorflow::io::BufferedInputStream* stream,
...
+      break;
+    }
+    // 2) find partial utf8, and read remain (< 4) bytes.
+    size_t remain = 0;
+    for (size_t i = 0; i < result_read.size(); i++) {
+      if (remain > 0) {
+        remain--;
+        continue;
+      }
+      if ((result_read[i] & 0xe0) == 0xc0) {
```
Comment (mihaimaruseac) [inline tensorflow/python/lib/io/file_io.i]: Let's extract the `0xe0`, `0xc0`, `0xf0` and `0xf8` magic constants to same named constants to increase readability

File: tensorflow/python/lib/io/file_io.i [hunk 3]
```diff
@@ -237,6 +237,67 @@ string ReadFromStream(tensorflow::io::BufferedInputStream* stream,
...
+        // n = 2
+        remain = 1;
+      }
+      else if ((result_read[i] & 0xf0) == 0xe0) {
```
Comment (mihaimaruseac) [inline tensorflow/python/lib/io/file_io.i]: Let's write these as

```
if (...) {
  ...
} else if (...) {
  ...
} else {
  ...
}
```

File: tensorflow/python/lib/io/file_io.i [hunk 4]
```diff
@@ -237,6 +237,67 @@ string ReadFromStream(tensorflow::io::BufferedInputStream* stream,
...
+                          TF_Status* status) {
+  tensorflow::tstring result;
+  size_t total = 0;
+  while (total < bytes) {
+    // Avoid over read:
+    // 1) read at least bytes (utf8 could be 1/2/3/4)
+    // 2) if string ends with partial utf8 (2/3/4), read remaining (< 4) bytes.
+    // 3) repeat until bytes == utf8 chars
+
+    // 1) read at least bytes - total
```
Comment (mihaimaruseac) [inline tensorflow/python/lib/io/file_io.i]: Let's move these to a comment in the header of the function, not in the `while` loop.

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]

**Comment by gbaned** [comment]
@yongtang  Can you please resolve conflicts? Thanks!

**Comment by yongtang** [comment]
Thanks @mihaimaruseac @gbaned for the review and sorry for the late response as the notice slipped through. The PR has been rebased and updated and the review comment has also been addressed. Please take a look.

**Review by mihaimaruseac — APPROVED** [review]

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]
This still seems to fail:

```
======================================================================
ERROR: testReadWriteNonBinaryModeUTF8 (__main__.FileIoTest)
testReadWriteNonBinaryModeUTF8 (__main__.FileIoTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/lib/io/file_io_test.runfiles/org_tensorflow/tensorflow/python/lib/io/file_io_test.py", line 686, in testReadWriteNonBinaryModeUTF8
    self.assertEqual(g.read(), f.read())
  File "/usr/lib/python3.6/encodings/ascii.py", line 26, in decode
    return codecs.ascii_decode(input, self.errors)[0]
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 1: ordinal not in range(128)
```

**Comment by gbaned** [comment]
@yongtang Can you please check @mihaimaruseac's comments and keep us posted ? Thanks!

**Comment by gbaned** [comment]
Hi @yongtang I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen! thank you for your contribution.

**Pull request closed without merging**